src/go/k8s/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -531,390 +524,294 @@ func setHelmLabelsAndAnnotations(object client.Object, rp *v1alpha2.Redpanda) {
-	}
-	if controllerutil.ContainsFinalizer(rp, FinalizerKey) {
-		controllerutil.RemoveFinalizer(rp, FinalizerKey)
-		if err := r.Client.Update(ctx, rp); err != nil {
-			return ctrl.Result{}, err
-		}
+		return err
 	}
-	return ctrl.Result{}, nil
-}
 
-func (r *RedpandaReconciler) createHelmRelease(ctx context.Context, rp *v1alpha2.Redpanda) (*helmv2beta2.HelmRelease, error) {
-	// create helmRelease resource from template
-	hRelease, err := r.createHelmReleaseFromTemplate(ctx, rp)
-	if err != nil {
-		r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, fmt.Sprintf("could not create helm release template: %s", err))
-		return hRelease, fmt.Errorf("could not create HelmRelease template: %w", err)
+	// do not apply objects that have not drifted to avoid bumping the resource version
+	if resource.GetResourceVersion() != "" &&
+		apiequality.Semantic.DeepEqual(dryRunObject.GetLabels(), existingObject.GetLabels()) &&
+		apiequality.Semantic.DeepEqual(dryRunObject.GetAnnotations(), existingObject.GetAnnotations()) &&
+		apiequality.Semantic.DeepEqual(prepareObjectForDiff(dryRunObject), prepareObjectForDiff(existingObject)) {
+		return nil
 	}
 
-	// create helmRelease object here
-	if err := r.Client.Create(ctx, hRelease); err != nil {
-		if !apierrors.IsAlreadyExists(err) {
-			r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, err.Error())
-			return hRelease, fmt.Errorf("failed to create HelmRelease '%s/%s': %w", rp.Namespace, rp.Status.HelmRelease, err)
-		}
-		// we already exist, then update the status to rp
-		rp.Status.HelmRelease = rp.GetHelmReleaseName()
-	}
+	appliedObject := resource.DeepCopy()
+	return r.apply(ctx, appliedObject)
+}
 
-	// we have created the resource, so we are ok to update events, and update the helmRelease name on the status object
-	r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityInfo, fmt.Sprintf("HelmRelease '%s/%s' created ", rp.Namespace, rp.GetHelmReleaseName()))
-	rp.Status.HelmRelease = rp.GetHelmReleaseName()
+// prepareObjectForDiff removes the metadata and status fields from the given object
+func prepareObjectForDiff(object *unstructured.Unstructured) *unstructured.Unstructured {
+	deepCopy := object.DeepCopy()
+	unstructured.RemoveNestedField(deepCopy.Object, "metadata")
+	unstructured.RemoveNestedField(deepCopy.Object, "status")
+	return deepCopy
+}
 
-	return hRelease, nil
+// Match CEL immutable error variants.
+var matchImmutableFieldErrors = []*regexp.Regexp{
+	regexp.MustCompile(`.*is\simmutable.*`),
+	regexp.MustCompile(`.*immutable\sfield.*`),
 }
 
-func (r *RedpandaReconciler) deleteHelmRelease(ctx context.Context, rp *v1alpha2.Redpanda) error {
-	if rp.Status.HelmRelease == "" {
-		return nil
+func isImmutableError(err error) bool {
+	// Detect immutability like kubectl does
+	// https://github.com/kubernetes/kubectl/blob/8165f83007/pkg/cmd/apply/patcher.go#L201
+	if k8serrors.IsConflict(err) || k8serrors.IsInvalid(err) {
```

Comment:
Is there any case aside from optimistic locking that would return a conflict error? AFAIK there isn't and the linked line does seem to only be checking for immutability.

Reply:
I'm not sure when `Conflict` error would be returned by K8S API server. I would say the same, that only optimistic locking can return such error.